extensions/amp-auto-ads/0.1/placement.js

```diff
@@ -265,24 +256,76 @@ function getPlacementFromObject(win, placementObj) {
       };
     }
   }
-  return new Placement(win, resourcesForDoc(anchorElement), anchorElement,
-      placementObj['pos'], injector, margins);
+  anchorElements.forEach(anchorElement => {
+    if ((placementObj['pos'] == Position.BEFORE ||
+         placementObj['pos'] == Position.AFTER) &&
+        !anchorElement.parentNode) {
+      dev().warn(TAG, 'Parentless anchor with BEFORE/AFTER position.');
+      return;
+    }
+    placements.push(new Placement(win, resourcesForDoc(anchorElement),
+        anchorElement, placementObj['pos'], injector, margins));
+  });
 }
 
 /**
+ * Looks up the element(s) addresses by the anchorObj.
+ *
+ * For a standard anchorObj this will return a single item array, or an empty
+ * array is no element on the page was addressed by the anchorObj.
+ *
+ * For a anchorObj containing a sub-anchor, sub-elements of the main selected
+ * element will be looked up and all matching elements returned. An example
+ * of how the sub-anchor feature might be used is looking up all <P> elements
+ * contained within the main content section of a page (where the main anchorObj
+ * selector identifies the main content section and the sub-anchor identifies a
+ * variable number of <P> elements).
+ *
  * @param {!Window} win
  * @param {!Object} anchorObj
- * @return {?Element}
+ * @return {!Array<!Element>}
  */
-function getAnchorElement(win, anchorObj) {
+function getAnchorElements(win, anchorObj) {
   const selector = anchorObj['selector'];
   if (!selector) {
     dev().warn(TAG, 'No selector in anchor');
-    return null;
+    return [];
   }
   const index = anchorObj['index'] || 0;
-  if (index == 0) {
-    return win.document.querySelector(selector);
+  const mainEl = index == 0 ?
+      win.document.querySelector(selector) :
+      win.document.querySelectorAll(selector)[index];
+  if (!mainEl) {
+    return [];
+  }
+  if (anchorObj['sub']) {
+    return getSubAnchorElements(mainEl, anchorObj['sub']);
   }
-  return win.document.querySelectorAll(selector)[index] || null;
+  return [mainEl];
+}
+
+/**
+ * A warning on sub-selectors: Calling Element.querySelectorAll limits the
+ * returned elements to ancestors of Element, but still calculates the CSS
+ * selector matching within the context of the whole DOM.
+ *
+ * @param {!Element} ancestorElement
+ * @param {!Object} subAnchor
+ * @return {!Array<!Element>}
+ */
+function getSubAnchorElements(ancestorElement, subAnchor) {
+  const selector = subAnchor['selector'];
+  if (!selector) {
+    dev().warn(TAG, 'No selector in sub-anchor');
+    return [];
+  }
+  const minChars = subAnchor['min_c'] || 0;
```

Comment:
There are potentially quite a few placements in the config, and since this is an over-the-wire format I think there is some value in trying to keep the keys inside the placement objects short, although gzipping maybe makes this less of a concern. I think explaining the config inside the .md file should hopefully be sufficient though. What is the preferred format for expressing a JSON schema in the amp project? I couldn't find any examples.

I've made the sub selector recursive, which also resolves the min_c being available to the main selector.